src/renderer/api/kube-object.ts

```diff
@@ -55,12 +56,233 @@ export class KubeObject implements ItemObject {
-    return !object.items && object.metadata;
+  static isJsonApiData(object: unknown): object is KubeJsonApiData {
+    if (!object || typeof object !== "object") {
+      return false;
+    }
+
+    if (!hasOwnProperties(object, "kind", "apiVersion", "metadata")) {
+      return false;
+    }
+
+    if (typeof object.kind !== "string"
+      || typeof object.apiVersion !== "string"
+      || !object.metadata || typeof object.metadata !== "object"
+    ) {
+      return false;
+    }
+
+    const { metadata } = object;
+
+    if (!hasOwnProperties(metadata, "uid", "name", "resourceVersion", "selfLink")) {
+      return false;
+    }
+
+    if (typeof metadata.uid !== "string"
+      || typeof metadata.name !== "string"
+      || typeof metadata.resourceVersion !== "string"
+      || typeof metadata.selfLink !== "string"
+    ) {
+      return false;
+    }
+
+    if (hasOwnProperty(metadata, "namespace")
+      && typeof metadata.namespace !== "string"
+      && typeof metadata.namespace !== "undefined"
+    ) {
+      return false;
+    }
+
+    if (hasOwnProperty(metadata, "creationTimestamp")
+      && typeof metadata.creationTimestamp !== "string"
+      && typeof metadata.creationTimestamp !== "undefined"
+    ) {
+      return false;
+    }
+
+    if (hasOwnProperty(metadata, "continue")
+      && typeof metadata.continue !== "string"
+      && typeof metadata.continue !== "undefined"
+    ) {
+      return false;
+    }
+
+    if (hasOwnProperty(metadata, "finalizers")
+      && typeof metadata.finalizers !== "undefined"
+      && !(Array.isArray(metadata.finalizers)
+      && metadata.finalizers.every(finalizer => typeof finalizer === "string"))
+    ) {
+      return false;
+    }
+
+    labels: if (hasOwnProperty(metadata, "labels")) {
+      if (typeof metadata.labels === "undefined" || (typeof metadata.labels === "object" && metadata.labels !== null && !Array.isArray(metadata.labels))) {
+        if (metadata.labels) {
+          for (const label in metadata.labels) {
+            if (!hasOwnProperty(metadata.labels, label) || typeof metadata.labels[label] !== "string") {
+              return false;
+            }
+          }
+
+          break labels;
+        }
+      }
+
+      return false;
+    }
+
+    annotations: if (hasOwnProperty(metadata, "annotations")) {
+      if (typeof metadata.annotations === "undefined" || (typeof metadata.annotations === "object" && metadata.annotations !== null && !Array.isArray(metadata.annotations))) {
+        if (metadata.annotations) {
+          for (const annotation in metadata.annotations) {
+            if (!hasOwnProperty(metadata.annotations, annotation) || typeof metadata.annotations[annotation] !== "string") {
+              return false;
+            }
+          }
+
+          break annotations;
+        }
+      }
+
+      return false;
+    }
+
+    return true;
+  }
+
+  static isPartialJsonApiData(object: unknown): object is Partial<KubeJsonApiData> {
```

Comment:
I don't get the difference between this and `isJsonApiData()`. Can duplicated code be shared? They look the same except for what `object` is.

Reply:
The difference is subtle yes, but then again so is the difference between `T` and `Partial<T>`. Namely, this function still returns true if any of the top level fields are `undefined` or not present. That is the only difference.

Unfortunately, there is no easy way to have a `isPartial` argument. However, I might be able to simplify the above by first calling this and then verifying that none of the fields are `undefined` or missing.